### DISCUSSION TITLE
Feature: types for open trees

### DISCUSSION THREAD

**Discussion by fmontesi** [jolie/jolie#420] [Ideas (Features, Fixes, etc.)]
Our types are roughly equivalent to "closed records": a type does not allow for any subnode that is not specified in the type.
The only way we have to allow for subnodes with unknown names is by using `{ ? }`, for example:
```jolie
type Hello { ? } // Hello can have any subnodes of any type
```

We could replace this with a more powerful feature: using `?` as a placeholder for "unknown names".

For example, the following type allows for any subnodes as long as they have type `int`:

```jolie
type Hello { ?:int }
```

This would allow us to specify some keys for which we want a different type, and then use `?` to have an "open record" (or map, or dictionary, if you like) where all the other nodes have some other type, as follows.

```jolie
type MyType { someNode:string, ?:SomeType } // someNode must be a string, all other nodes must be of type `SomeType` (user-defined, here left unspecified)
```

The old `?` primitive would become just syntactic sugar for this new primitive: `type T { ? }` would be interpreted as `type T { ?:undefined }`.

With this primitive, we could actually write the `undefined` type in Jolie as the following recursive type:
```jolie
type undefined:any { ?:undefined }
```

**Comment by klag** [comment]
Great!

what about giving also the possibility to express a regex?

**Reply by fmontesi** [reply]
Good question. I'm concerned with how this might impact the computation of type operations, but I *guess* (not sure) that there is hope if we keep to a regex language with decidable inclusion/comparison operations like we use in our refinement types.
Do you have a proposal for a syntax? Preferably one that would allow us to implement the easy `?` first (as in my opening post) and then implement improvements (like your regexes) in a later version?

Something like `?(REGEX):T`?

**Reply by klag** [reply]
we could use the same syntax of refinement types:

`regex(".*@.*\\..*"): T`

**Reply by thesave** [reply]
Nice proposal :)

Regarding the regexp types, just out of curiosity, @klag, do you have some examples (like, that you met in production) where you needed regular expressions for types? 
I guess it could be useful to check structures like 

```json
{
 "title": "My Movie"
 "metadata": {
    "movie-date" : "..."
    "movie-release" : "..."
    "movie-rating" : "..."
  }
}
```

where, e.g.,  you want to allow only for an open-ended set of leaves with the "movie-" prefix under the node "metadata".
Is that the case? Are there more complex examples?
